Add coinshuffle-plusplus-part-2 review to wasabi-research-club (#642)
bitcointranscripts/bitcointranscripts: wasabi-research-club/coinshuffle-plusplus-part-2.md

Reviewer: ```suggestion
And in this talk, our peer-to-peer mixing protocol is CoinShuffle++.
```

Author: just curious, are you using AI for reviews ?

Reviewer: @Garvit-77 
Yes, I'm using AI, not the bot way. Still, I need to manually double-check and research. If I do a thorough sweep, I might be able to find more - especially videos with multiple people speaking.

Manual check:
https://github.com/bitcointranscripts/bitcointranscripts/pull/642#discussion_r3162755314
Cross-check:
https://github.com/bitcointranscripts/bitcointranscripts/pull/635#discussion_r3149035792

I wasn't asked to watch the entire video right now. If I had to watch, research, and rewatch it all, it would take several times as long. Since you’re watching it from start to finish, please let me know if you disagree with any of my review comments; you might be right!

Author: Yeahh well, I was just curious how and what's your workflow

Reviewer: ```suggestion
And then there's something in the middle, maybe, that I can briefly mention in the end here, which is ValueShuffle.
```

aliases:
  - '/wasabi/research-club/coinshuffle-plusplus-part-2/'
---
**Anchor:** 00:00:02

All right guys, we're here with Tim Ruffing, one of the authors of the `CoinShuffle++` paper to talk about privacy.
So this PowerPoint and the presentation, Tim will do that himself and I'll just sit back.

> **TIm Ruffing:** 00:00:22

Reviewer: ```suggestion
**Tim Ruffing:** 00:00:22
```

Reviewer: Please double-check throughout the entire transcript. Same misspelled errors

## Big picture of privacy technologies.

**TIm Ruffing:** 00:02:45

So, improving privacy.
And here I have some picture with some privacy technologies.
This is just to show you where where CoinShuffle stands in this big picture.
In this big picture.
So, actually I could basically put CoinJoin instead here probably and it would still be a valid picture.
The basic idea here is that with CoinJoin or CoinShuffle or CoinMixing in general, also TumbleBit is in the same category, we can't do so much in terms of privacy, but we are pretty compatible with Bitcoin because it works currently in Bitcoin.

> Of course, if you want more privacy, you can have fancy zero-knowledge proofs and other technologies like zero-cash or whatever.

Reviewer: ```suggestion
Of course, if you want more privacy, you can have fancy zero-knowledge proofs and other technologies like Zerocash or whatever.
```

Reviewer: http://zerocash-project.org/how_zerocash_works

Author: thanks for this

## Coin mixing.

**TIm Ruffing:** 00:04:19

> Okay, this is a coin join seminar.

Reviewer: ```suggestion
Okay, this is a CoinJoin seminar.
```

And just in information, theoretically, this, I mean, this is not a proper explanation, but at least if you look at the number of bits here, that makes sense, right?
We have n messages altogether because we have n users.
And let's say every message has b bits.
So, if you have the list of the power sums here, we have n times b bits and n times b bits is is the same amount of information and because we used to use a proper encoding like this or some thing is just one encoding of a list of messages you can decode it back and get the messages back but for this to work we need finite field arithmetic and that's why we do sums and finite fields instead of just XOR.
This wouldn't work if you used XOR.

## Disruption.

**Tim Ruffing:** 00:27:15

Now the big problem in DC-net is that they can be disrupted, which basically means if there's one malicious user, let's say your Bob is malicious, what Bob could do is instead of sending m2 and m2^2 and so on and following this nice algebraic structure that you're supposed to follow, Bob could just send bullshit in every slot, for example, like random values or anything else that he wants, right?

> And now the problem is if we now want to sum up in the slots, of course we can sum up, we get some sums, but like those sums will be again bullshit, even worse, because we are doing like n DC-net here, or a DC-net in every of the n slots.

Reviewer: ```suggestion
And now the problem is if we now want to sum up in the slots, of course we can sum up, we get some sums, but like those sums will be again bullshit, even worse, because we are doing like n DC-nets here, or a DC-net in every of the n slots.
```

Reviewer: <img width="1652" height="290" alt="image" src="https://github.com/user-attachments/assets/5ee3c1d3-eed8-4d42-832c-870c3c761ede" />

Reviewer: <img width="1777" height="1089" alt="image" src="https://github.com/user-attachments/assets/df3dc9cf-f0ce-4e23-a50f-9fa90b663b8f" />

Reviewer: <img width="1649" height="688" alt="image" src="https://github.com/user-attachments/assets/4400f260-4a29-4626-9b61-60b9336fd379" />

Reviewer: Would you mind verifying the last [commit](https://github.com/bitcointranscripts/bitcointranscripts/pull/642/changes/7cc5d5841183e1a1baf5dcc16735cc58ed5fee50)? It looks like the change might have been applied to the wrong spot. Thanks

What I mean by that is we don't rely on anything like Tor to provide anonymity.
So in this example here now, look at the left-hand side.
Alice and Dave could be malicious and this means for anonymity that the anonymity set remains the set of honest users.
So Bob and Carol still have an anonymity among each other, Off course now they only have an anonymity set of two because they can't have anonymity together with the attacker, right?
Just not possible.
Because exactly, we have those links here.
So if the attacker controls Alice, then the attacker knows that A' is Alice's address and the attacker controls Dave, then he knows that D' is Dave's address.
So yeah, we can't provide anonymity with those two addresses.
But as I said, like Bob and Carol still get an anonymity set of two in this example.

## Termination in the presence of malicious users.

**TIm Ruffing:** 00:07:44

Another property that we want to have next to anonymity is termination.
This simply means the protocol terminates in the presence of malicious users, which means that there shouldn't be users that can stop honest users from finishing the protocol.

> The only thing we assume here is what I mentioned last week in the informal discussion about Crunchyroll++ already.

Reviewer: ```suggestion
The only thing we assume here is what I mentioned last week in the informal discussion about CoinShuffle++ already.
```

For termination, we assume that there is a bonded board.
What do we mean by that?
It's basically a server in the middle where we all connect to, and the server handles the broadcast for us and because we are in this peer-to-peer trust model, we don't really trust the server for anonymity, but we trust it for termination.
In practice, this means that if the server is malicious or just broken, then the peers won't be able to finish the protocol, but nothing bad happens.
They could just switch to a different server.
However, no matter how malicious the server is, the server can't break anonymity.

## How does `CoinShuffle++` work? / DC-net (Chaum, CRYPTO '88).

**TIm Ruffing:** 00:09:00

So how does `CoinShuffle++` actually work?

> It's based on a DC mode which basically stands for Dining Cryptographers Network.

Reviewer: <img height="300" alt="Image" src="https://github.com/user-attachments/assets/b8119d87-94f0-43bd-8074-296b95a89e8c" />

> It's based on a DC mode which basically stands for Dining Cryptographers Network.

Reviewer: ```suggestion
It's based on a DC-net which basically stands for Dining Cryptographers Network.
```

And now, the cool thing now is if we add those three messages up that have been broadcast, which is (1 + 1 + 0), then this basically means that we have...
Yeah, so if I expand it, we have just this result and now we see that every key has been added twice, which is kind of clear because it has been added by both sides of the of every of those connections basically.
So that means that all the keys here cancel out and we are left with the sum of the messages here, which is (1 + 1 + 0).
Yes, and the last 0 should be bold.
So If we add those together, we get 0 and 0 is indeed the sum of the messages that the users had in mind.

## DC-nets in practice.

**TIm Ruffing:** 00:12:08

So the interesting thing here is now that this gives the users some form of anonymity in this very simple setting.
So, now we know what the sum of the input messages is, but we don't know who contributed what part to this sum.
This is a form of anonymity, basically.
This was kind of a very simple and awkward example because if you want to do this in practice, we need to obtain shared symmetric keys.
This is not hard because we can just do a cryptographic key exchange.
This is kind of standard.

> We can do the key addition, for example.

Reviewer: ```suggestion
We can do the Diffie-Hellman key exchange, for example.
```

> But functionality-wise, What we want to do in practice is we don't want to send just one bit message, we want to send longer messages, for example, with common choices.

Reviewer: ```suggestion
But functionality-wise, What we want to do in practice is we don't want to send just one bit message, we want to send longer messages, for example, with Bitcoin addresses.
```

So that means that all the keys here cancel out and we are left with the sum of the messages here, which is (1 + 1 + 0).
Yes, and the last 0 should be bold.
So If we add those together, we get 0 and 0 is indeed the sum of the messages that the users had in mind.

## DC-nets in practice.

**Tim Ruffing:** 00:12:08

So the interesting thing here is now that this gives the users some form of anonymity in this very simple setting.
So, now we know what the sum of the input messages is, but we don't know who contributed what part to this sum.
This is a form of anonymity, basically.
This was kind of a very simple and awkward example because if you want to do this in practice, we need to obtain shared symmetric keys.
This is not hard because we can just do a cryptographic key exchange.
This is kind of standard.
We can do the Diffie-Hellman key exchange, for example.
But functionality-wise, What we want to do in practice is we don't want to send just one bit message, we want to send longer messages, for example, with Bitcoin addresses.
This is also not too hard to do.
Now I've shown you an example with the bit field, which is the field which is 0 and 1 basically as a mathematical field.

> What we can also do if you want to send larger messages is encode them as larger integers and use larger finite fields to transmute them.

Reviewer: ```suggestion
What we can also do if you want to send larger messages is encode them as larger integers and use larger finite fields to transmit them.
```

Three slots and next.
And then Alice in the slots where people don't have assignments, they just send zeros, right?
So if you look at the first slot, we add the zeros which are basically just some form of padding, I could say, and Bob's message.
Then we would have the messages.

## Slot assignment.

**TIm Ruffing:** 00:15:49

We could have the messages back, but the problem here is that this needs this, what I call magic now, this needs an anonymous slot assignment.
So in order to make this work, we already need some protocol that provides anonymity in some sense and this is where most of these proposals to use slot reservation actually fail in practice, because it's a chicken and egg problem.
You need, in order to get an anonymity here, you need another anonymous sub-protocol that gives you a slot assignment, because this is hard, those protocols in the literature do the various tricks that mostly are not great because they can fail in practice.
So for example, what they often do is that you guess a slot randomly.
Say we have a thousand slots available and there are users, and then every user guesses a random slot.

> And this kind of works in practice, but it's pretty annoying because now instead of running 3 ADC nets, we need to run a thousand 3C nets.

Reviewer: ```suggestion
And this kind of works in practice, but it's pretty annoying because now instead of running 3 DC-nets, we need to run a thousand DC-nets.
```

## Alphabetic ordering of public keys for slot assignment.

**TIm Ruffing:** 00:17:46

What we do instead is based on an idea that's pretty old already.

> It appeared shortly after the CEMENTS.

Reviewer: ```suggestion
It appeared shortly after the DC-nets.
```

**Tim Ruffing:** 00:27:15

Now the big problem in DCNets is that they can be disrupted, which basically means if there's one malicious user, let's say your Bob is malicious, what Bob could do is instead of sending m2 and m2^2 and so on and following this nice algebraic structure that you're supposed to follow, Bob could just send bullshit in every slot, for example, like random values or anything else that he wants, right?

> And now the problem is if we now want to sum up in the slots, offcourse we can sum up, we get some sums, but like those sums will be again bullshit, even worse, because we are doing like n DCNets here, or a DCNet in every of the n slots.

Reviewer: Could you verify all the `offcourse` changes in the last commit [549bda1](https://github.com/bitcointranscripts/bitcointranscripts/pull/642/commits/549bda1fd582be1ca875680ad358b1aa990fb19d)? 
00:28:10

Author: Doesn't this sounds right? grammatically

Reviewer: https://www.merriam-webster.com/dictionary/of%20course

Could you google or look up if `offcourse` exists in corpus or dictionary?

So What do we do then?
Well, in case of disruption, break anonymity.

## Why `CoinShuffle++` works. / Flowchart of CoinShuffle

**TIm Ruffing:** 00:29:07

And, well, this may sound like a bad idea, but let me explain why this actually works.
And to understand why this makes sense, we need to have a look at the flowchart of the coin shuffle run.

> So the first thing we do, like everybody generates the fresh Bitcoin interest, this is gonna be the output in the coin transaction on the right side and it will be our message in the peer-to-peer mixing protocol.

Reviewer: ```suggestion
So the first thing we do, like everybody generates the fresh Bitcoin address, this is gonna be the output in the coin transaction on the right side and it will be our message in the peer-to-peer mixing protocol.
```

Then, peers do key exchanges.

> They run the Hellman key exchanger.

Reviewer: ```suggestion
They run the Diffie-Hellman key exchanger.
```

I don't need to explain you how it works.
Just it makes sure that every pair of users will have a shared key that they need for the DCnet.
Okay, then we run the actual DC net and then we would run it in slots as I've just shown you.

> And then now If we would proceed like I've shown you previously, and actually this will not be the final flowchart of CronShuffle++, We will come to that later, but bear with me for the moment.

Reviewer: ```suggestion
And then now If we would proceed like I've shown you previously, and actually this will not be the final flowchart of CoinShuffle++, We will come to that later, but bear with me for the moment.
```

And then he could, it's not clear from the slide here but then you could construct a special form of bullshit message in a sense that only disrupts for example Alice's message and not the other messages.
Okay?

**Audience 1:** 00:33:56

No, not okay, because then everyone would be able to check if Alice's message is a Bitcoin address, valid Bitcoin address or not.

**TIm Ruffing:** 00:34:08

So, what Bob sees is, Bob sees just a list of the other messages.
He doesn't know which message belongs to which user.
So he sees somehow that M1 is there, which is Alice's message, but he doesn't know that it belongs to Alice.
But now, because he knows M1, he could send bullshit such that m1 is disrupted, is basically replaced by a random message or a message of his choice even and the other messages are not.

**Audience 1:** 00:34:40

Yes

**TIm Ruffing:** 00:34:41

Okay.
So, he could selectively disrupt messages even though he doesn't know to which users those messages belong.

**Audience 1:** 00:34:53

Makes sense thank you.
So he could swap Alice's address, Swap a random address to his own address

**TIm Ruffing:** 00:35:02

> For example, yeah, What I was saying before is that the only thing Alice now at the end of the DZNet round, what she can do is she can look at the set of messages of the results of the protocol and see if her old message is there, right?

Reviewer: ```suggestion
For example, yeah, What I was saying before is that the only thing Alice now at the end of the DC-net round, what she can do is she can look at the set of messages of the results of the protocol and see if her old message is there, right?
```

But she doesn't know if the other messages are all there.
She will see some messages there, for example, if Kevil's message has been replaced or not.

> And the important thing at this step here is that, however, all the peers in the protocol need to agree whether disruption has happened or not because if it has not if there was no disruption then like this is the case we see now on the slides and then they will go ahead and create a contract transaction and hopefully sign it.

Reviewer: ```suggestion
And the important thing at this step here is that, however, all the peers in the protocol need to agree whether disruption has happened or not because if it has not if there was no disruption then like this is the case we see now on the slides and then they will go ahead and create a CoinJoin transaction and hopefully sign it.
```

> However, if one of them was disrupted, then what they want to do is They want to reveal the key exchange sequence, which is basically a standard de-anonymization step.

Reviewer: ```suggestion
However, if one of them was disrupted, then what they want to do is They want to reveal the key exchange secrets, which is basically a standard de-anonymization step.
```

Then we can ask is there a protocol in the intersection of all these circuits in the middle.

## Define termination.

**Audience 1:** 00:43:20

what the termination mean

**Tim Ruffing:** 00:43:26

Termination means that the protocol terminates, even if there are malicious peers inside.

**Audience 1:** 00:43:34

So not termination would mean if there are malicious peers then the protocol just doesn't.

**Tim Ruffing:** 00:43:46

It doesn't finish.
So, or it aborts.

**Audience 1:** 00:43:55

So it's like termination is?

**Tim Ruffing:** 00:43:59

Maybe it should have been called successful termination, right?
So you can imagine that a protocol that doesn't provide termination basically, for example, it tries to start mixing with the DCnet and then it notices, oh, there was disruption and then, well, it fails, just aborts.

**Audience 1:** 00:44:21

Okay, so if it aborts, if it notices that there was disruption, then that would mean termination is not ensured.

**Tim Ruffing:** 00:44:31

Right.

## Wasabi does not guarantee termination.

**Anchor:** 00:44:35

Yeah, so I think Adam, like currently Wasabi does not guarantee termination, right?

> Because if a `CoinJoin` doesn't happen, wasabi bans a coin and then reopens the round and more malicious people could enter the round.

Reviewer: ```suggestion
Because if a `CoinJoin` doesn't happen, Wasabi bans a coin and then reopens the round and more malicious people could enter the round.
```

Does this answer your question or not?

**Audience 1:** 00:48:45

No, my point was that if you can identify the malicious peer, then it doesn't really matter if termination happens or not of this protocol rounds.

**TIm Ruffing:** 00:49:11

Okay, again, when I say termination, I don't mean termination of a protocol run, of a run that can be aborted.
I mean termination of the entire thing, like the entire thing with a big loop.

**Audience 1:** 00:49:33

So, if you can identify the malicious peer, but you don't exclude it, but you have to create a completely new protocol and from that you exclude it that would mean that it is a terminating protocol.
Okay fine, let's go.

**TIm Ruffing:** 00:50:00

Okay.

## `CoinShuffle++` is a circular protocol.

**Anchor:** 00:50:04

Just to clarify, `CoinShuffle++` is a circular protocol.
It has six steps, and then four of them are repeated every time someone disrupts, right?

**TIm Ruffing:** 00:50:19

And that's why I look at the entire thing, like with the circular thing in It provides termination.
I think this is all I wanted to say here.

## Anonymity attack on Dissent Protocol.

**TIm Ruffing:** 00:50:39

> I can actually show you how this anonymity attack on this end works.

Reviewer: ```suggestion
I can actually show you how this anonymity attack on Dissent works.
```

> I don't need to tell you a lot about this end but I think it can still be instructive.

Reviewer: ```suggestion
I don't need to tell you a lot about Dissent but I think it can still be instructive.
```

Reviewer: 50:49

> The only thing I think you need to know is that like the three points on top here, this end proceeds in broadcast rounds and the outcome of the protocol is revealed to all the users in the last broadcast and also a network attacker sees the outcome of the protocol in the last broadcast.

Reviewer: ```suggestion
The only thing I think you need to know is that like the three points on top here, Dissent proceeds in broadcast rounds and the outcome of the protocol is revealed to all the users in the last broadcast and also a network attacker sees the outcome of the protocol in the last broadcast.
```

Because he knows that he blocked Bob, he knows that this is the message from Bob.
Does this make sense?

**Audience 1:** 00:54:03

Probably.

**Tim Ruffing:** 00:54:05

Okay, feel free to ask.

**Audience 1:** 00:54:09

Sorry I was just blanking out.
It was really a good explanation.

**Anchor:** 00:54:14

Okay.

> So, I guess I just don't understand how descent works well enough, like why there would be multiple rounds.

Reviewer: ```suggestion
So, I guess I just don't understand how Dissent works well enough, like why there would be multiple rounds.
```

> So, I guess I just don't understand how dissent works well enough, like why there would be multiple rounds.

Reviewer: ```suggestion
So, I guess I just don't understand how Dissent works well enough, like why there would be multiple rounds.
```

But isn't it the case that Bob's message is XOR with the two adjacent Diffie-Hellman shared secrets?
So How is it that you can figure out Bob's message?

**TIm Ruffing:** 00:54:43

Okay, I think there are multiple things here.

> So, yes, I didn't explain to you how this works.

Reviewer: ```suggestion
So, yes, I didn't explain to you how Dissent works.
```

> It works differently from a DCnet.

Reviewer: ```suggestion
It works differently from a DC-net.
```

> So basically, it's a little bit like or it's pretty similar to the contra of the one that you also discussed.

Reviewer: ```suggestion
So basically, it's a little bit like or it's pretty similar to the CoinJoin of the one that you also discussed.
```

But in the end of the protocol, at some point you need to reveal the messages, right?
So what the attacker here does is in the first run, like he just looks at the protocol passively until the last round.
So basically he observes all the messages of all the people and then she just he just sees the outcome of the protocol which is the list of messages and so far there's nothing wrong with it because he doesn't learn which message belongs to the chooser

## How practical is it to block an adversary from the protocol.

**Anchor:** 00:56:14

This might be a dumb question but how practical is it to block someone from the protocol, mid protocol?

**Tim Ruffing:** 00:56:23

I think this really depends on the network setting.
Like usually if you write protocols, you want them to be secure against network attackers and we actually assume that network attackers can do much more.
They can not only block messages, they can also replace them, duplicate them and all the stuff.
Of course you can ask how realistic that in practice is.
Usually we abstract away from this question, just say like look the attacker is super powerful because we don't know how powerful he actually is.

> Now, if you look at `CoinShuffle` or descent specifically, I told you in the beginning that we assume that we have this bulletin board in the middle that handles all broadcasts.

Reviewer: ```suggestion
Now, if you look at `CoinShuffle` or Dissent specifically, I told you in the beginning that we assume that we have this bulletin board in the middle that handles all broadcasts.
```

And the reason why we do this actually is to have better efficiency.
And Now it's getting interesting.
So because we didn't want to trust this bulletin board in the middle of the server in the middle for anonymity.
But if you think of the server in the middle being the malicious network attacker, Then suddenly it's very, very easy to block messages, right?
You just don't forward them.

> Like think of, I think last week I mentioned the simple example of my RST server, just as an example of a server that broadcasts your messages to everybody else.

Reviewer: ```suggestion
Like think of, I think last week I mentioned the simple example of my IRC server, just as an example of a server that broadcasts your messages to everybody else.
```

Okay, let's count communication rounds.
And communication rounds are here, broadcasts.
So we need one broadcast for the key exchange.
Then we need one broadcast to run the actual DCNets.
And now, because now is the crucial part, because we all need to, like all the peers now need to agree on whether the protocol has been disrupted or not.
I told you this example where you can selectively disrupt messages from peers.
So to avoid this, in a naive way, we would need another broadcast here just to check that we all agree on whether there was disruption or not because as I explained to you earlier the only thing you can do is you can look at this on the output of the protocol and see if his own message is there but he doesn't know if the messages of the others are there.
So to make sure that we all agree whether there was disruption or not, we would need another broadcast from here.
And this is the broadcast that basically belongs to the disruption box.

> And then once we all know whether there has been disruption or not, we need one more broadcast either to broadcast the signatures on the contract transaction or to reveal the key exchange secrets depending on which way we go now.

Reviewer: ```suggestion
And then once we all know whether there has been disruption or not, we need one more broadcast either to broadcast the signatures on the CoinJoin transaction or to reveal the key exchange secrets depending on which way we go now.
```

Does this make sense?
Please ask if this is unclear.

**Anchor:** 01:07:42

Makes sense to me.

**TIm Ruffing:** 01:07:45

Great.
And this is the reason why we wanted to push the DCNET round to the third broadcast instead of the second.
So This was the thing I've shown you on the previous slides, right?
In the naive version, we run key exchange, then already DCNET, and then we need to check with the others if disruption happened.
Then the DCnet round would be in the, the DCnet message would be in the second round.
But here what we act, what we here do is to this commitments to the DCnet And then sending the DCnet only in the third round, we move the DCnet to the third round and we make this interleaving of two here possible.
Yeah, and now if you count the number of rounds, if you have F malicious users, then we need 4 plus 2 F broadcast rounds.

> And this is better than previous work, which was the original cold travel, which was just all of NF rounds.

Reviewer: ```suggestion
And this is better than previous work, which was the original CoinShuffle, which was just all of NF rounds.
```

> And this is better than previous work, which was the original CoinShuffle, which was just all of NF rounds.

Reviewer: Could you watch the video and double-check "...all of NF rounds"? It looks like a phonetic transcription of Big O notation

Reviewer: <img width="1964" height="1425" alt="image" src="https://github.com/user-attachments/assets/64cf0eaa-2fa3-406f-ab1d-f66028b9a5e6" />

So much slower.

## Practical evaluation of `CoinShuffle++`.

**Tim Ruffing:** 01:09:04

> Yeah, and now we can evaluate this and practice with the with the basically the setup that I mentioned here on the slide and compare this to the to the previous protocol coin shuffle And at least in this setting, `CoinShuffle++` is much, much better than Coinshuffle, as you can see on the graph.

Reviewer: ```suggestion
Yeah, and now we can evaluate this and practice with the with the basically the setup that I mentioned here on the slide and compare this to the to the previous protocol CoinShuffle And at least in this setting, `CoinShuffle++` is much, much better than CoinShuffle, as you can see on the graph.
```

So Yeah, for example, for usually a number of you point out on the paper is 50 notes there.
We are still below 10 seconds in this setting.

> And then coin shuffle, the original coin shuffle was like almost three minutes in the setting.

Reviewer: ```suggestion
And then CoinShuffle, the original CoinShuffle was like almost three minutes in the setting.
```

## Handling unequal inputs.

**Tim Ruffing:** 01:10:00

And then just, I think this is maybe should have been called limitations of CoinJoin actually for this audience here.
But you're probably aware that handling unequal inputs is a very naive form, is a very bad idea, right?
Because if you do a transaction like it's on the slide here, then everybody could just, every external observer even could just tell that Bob, like the one point.
Yeah, so if everybody who looks at this transaction can just tell that B and B prime belong together, there's no privacy if you do a naive CoinJoin with unequal inputs.
And we all know Adam has much better suggestions to do this.

> But this is still an obvious limitation of coin drawings.

Reviewer: ```suggestion
But this is still an obvious limitation of CoinJoins.
```

Reviewer: Could you check this one? Is it `drawings` , instead?

It's nothing that you can discard and throw a new random one, right?
It just doesn't work.

**Audience 1:** 01:14:02

Because does that matter?
What really matters is that the output address is changed.
So it's like the message is not the 0.1.
The message is or 0.1.
If you next you message or to 0.1 because Robert gave you many addresses.

**TIm Ruffing:** 01:14:27

Yeah, but I think like I had this we had this side topic, right?
Is a fresh message is actually necessary?
Can we do mixed, sorry, can we do fixed messages?
And the reason why the answer was no, the reason was that anonymity is broken.
So functionally you're right, you can just like if the first run is disrupted you could abort it and then restart with R2 and 0.1. But then it could be that there is an attacker and anonymity.

**Audience 1:** 01:15:13

I'm sorry, can you repeat that?

## Mixing and paying simultaneously.

**TIm Ruffing:** 01:15:22

I had a few slides under the title, side topic are fresh messages necessary, right?

> Where I showed the attack on this end, for example.

Reviewer: ```suggestion
Where I showed the attack on Dissent, for example.
```

**Audience 1:** 01:15:38

But the message is fresh.
It's just one component of the message is fixed.

**Tim Ruffing:** 01:15:44

Okay.
But the attack basically applies to also messages where one component is fixed.
You really lose anonymity in this setting.

**Anchor:** 01:16:00

Also how would you pay someone, like typically you only have one address?

**Tim Ruffing:** 01:16:08

Can you repeat?

**Anchor:** 01:16:12

Typically you only get one address when you have to pay someone.

**Tim Ruffing:** 01:16:17

Yeah, this is what I like.
If I write a paper, I would call this an engineering challenge and put it aside.
But you're right.
And so what I'm saying is you're right.
In practice, you need to solve the problem.
It's just that it's kind of doable, but you need recipient support and this would be another drawback, even if it would be possible.

> Either, like if you do a coinjoin with 50 parties, then either you ask the pizza restaurant to give you 49 addresses, which is ugly but works.

Reviewer: ```suggestion
Either, like if you do a CoinJoin with 50 parties, then either you ask the pizza restaurant to give you 49 addresses, which is ugly but works.
```

**Audience 1:** 01:18:38

Yes.

**TIm Ruffing:** 01:18:40

Yeah, this, you're right, this principle would work.
It's a good question how robust this is in practice, but in principle you could do this.

**Audience 1:** 01:18:54

All right, thank you.

**TIm Ruffing:** 01:18:56

Okay, and now this is the final word here on the slides, I think.
So, I hinted that this already, Like if we are in a system where we have confidential transactions, which is a cryptographic technique to hide the amounts on the transactions, like the amounts are not in plain, stored in plain, but are in homomorphic commitments and cryptographic commitments.
Then suddenly those amounts, because they are in commitments, These are not fixed messages anymore.
This can be re-randomized basically.
So you can take a commitment and like you can commit to the same amount twice and those commitments look totally independent of each other.
Then this problem with the fixed message goes away and then you suddenly can mix and pay simultaneously.

> And also like an attentive protocol you need for this is value shuffle, which is full `OP_fork`.

Reviewer: There is no such opcode as "OP_fork"

> And also like an attentive protocol you need for this is value shuffle, which is full `OP_fork`.

Reviewer: ```suggestion
And also like an attentive protocol you need for this is ValueShuffle, which is full of work.
```

Reviewer: @0tuedon 
I'm not sure about this one. It sounds like "`full of work`" to me
https://www.youtube.com/watch?v=bCZFAqB3bnU&t=4800s

Reviewer: ValueShuffle  
https://bitcoinmagazine.com/culture/valueshuffle-brings-together-the-best-of-both-worlds-for-privacy-1483557170

Reviewer 2: > https://bitcoinmagazine.com/culture/valueshuffle-brings-together-the-best-of-both-worlds-for-privacy-1483557170

I doubt there is an op_fork lol I heard "full of work"

Reviewer: @Garvit-77 You did not fix "value shuffle" to "ValueShuffle". 

Why is this resolved?

> And also like an attentive protocol you need for this is value shuffle, which is full of work.

Reviewer: ```suggestion
And also like an attentive protocol you need for this is ValueShuffle, which is full of work.
```

Reviewer: Please don't **resolve** the thread if suggested change is not fixed.
I cannot unresolve a thread. I have to "show resolved" to find it. 

You did not change "value shuffle" to "ValueShuffle".

<img width="1682" height="1604" alt="image" src="https://github.com/user-attachments/assets/c3b36abe-3fcb-4dcb-814e-bb87fcaf647c" />

Author: thank you for acknowledging that, but i guess i had resolved it after change still I would keep it in check further.

And also you, of course, get rid of the previous problem that you can't mix unequal amounts, then you can also mix equal amounts and suddenly mixing is so much nicer and better.
If you could hide the amounts, but I think this is also not used in this seminar here.

**Audience 1:** 01:20:32

I have one more theoretical and less practical talk that you know what does termination mean here?
Termination means that everyone signs, but that's not really what termination is ?
Termination is that successful broadcast happens.
So now there is the issue with double spends.

**TIm Ruffing:** 01:21:01

Yeah, right.

**Audience 1:** 01:21:02

Yeah, so we detect the double spends and then we have to run the whole protocol again.
What if you are a miner then everyone has to be online until confirmation.

**Anchor:** 01:21:18

Sorry but isn't a double spend the same as not signing the CoinJoin at the end round?

**TIm Ruffing:** 01:21:25

The problem is just that you don't figure it out so quickly, right?
So and you don't know, right?

> As I've shown you on the slides, basically, the protocol is finished as soon as you get the signatures on the CoinJoin message, the content of the action from everybody else, then you have a signed CoinJoin.

Reviewer: ```suggestion
As I've shown you on the slides, basically, the protocol is finished as soon as you get the signatures on the CoinJoin message, the CoinJoin transaction from everybody else, then you have a signed CoinJoin.
```

Yeah, it's like, other schemes those are working in practice don't even ensure denial of service protection right and this is like a new level of theoretical attacks.
So it's an engineering challenge, as you would put it.

**Tim Ruffing:** 01:24:23

Yeah, I mean, I would at least say it's not the end of the world.
But like, Even if people do this attack actively, what you can do is wait for another minute or so.
If you haven't seen a conflicting transaction for a minute, well, okay.
And actually, the peers can also then relay the transactions, right?
Like if one of the peers on the protocol sees a conflicting transaction, you could relay it to the others and basically have proof that there's a double-span.

**Audience 1:** 01:25:05

My point would be, but you wouldn't see because the nodes don't broadcast to you, but yeah, you can relay to the others.

**Tim Ruffing:** 01:25:13

But it's kind of, from a running time perspective, kind of annoying.
Right now I told you, okay, like in a nice network setting, this protocol takes below 10 seconds, and then I tell you, no, you need to wait another 20 seconds just to make sure there's no double spam.

**Audience 1:** 01:25:33

> I mean if others relay a transaction like that, a conflicting transaction, then would that, I mean you would have to trust the other peers that they are not lying about the conflict in terms oh no because they can only the transaction those are theirs so they would be the the the the guys who are disrupting around so yeah never mind.

Reviewer: ```suggestion
I mean if others relay a transaction like that, a conflicting transaction, then would that, I mean you would have to trust the other peers that they are not lying about the conflicting transaction oh no because they can only the transaction those are theirs so they would be the the the the guys who are disrupting around so yeah never mind.
```
